### ISSUE TITLE
Missing parameters to reproduce ROC curves

### ISSUE THREAD

**Issue by johannesjung** [lsgos/uncertainty-adversarial-paper#2]
Hi, 
thank you for making the code needed to reproduce the experiments from your paper publicly available. Currently, I'm trying to reproduce the ROC curves in figure 14 from the paper. Apparently the default parameters of this code version produce very different results, some examples attached below.

Could you please share the necessary parameter configuration for

- number of examples
- number of MC passes
- exact configuration for the attack parameters
- seed for the random number generator to select the same images

Training the model worked fine. The validation accuracy after 40 epochs was about 94%. However, it would also be nice if you could share your model parameters to rule out any doubts in this regard.

I just added a title to the plot to identify the related attack configuration. I haven't changed any of the code's logic.
![image](https://user-images.githubusercontent.com/24568449/39998150-7f49fca0-5785-11e8-81ed-beab6ba5c63e.png)
![image](https://user-images.githubusercontent.com/24568449/39997998-f622e19e-5784-11e8-9cd3-7d81f551f512.png)

**Comment by lsgos** [comment]
Hmm that’s a bit worrying; though those fgm results are more or less what we got in the paper if I’m not mistaken? The momentum method is surprising though...

I think I used around 200 examples. 
You need to use about 20 (at least 10) or so mc samples , otherwise the estimates aren’t very good. The amount you might be able to actually use may be limited by the gpu you have because the models are so big.  How many did you use to generate these plots?

The exact configurations for the attacks we used are the ones in this codebase, so those should be fine. 

I didn’t fix the random seeds, unfortunately (as you can see from the code). There shouldn’t be this much variation though?

Let me know how you get on; if you still have problems I can maybe re-run the experiments on my end too (though I’m a little busy with other work at the moment). 

If I had to guess, I’d say you need to use more samples in the forward pass; since the mc model and deterministic model use the same weights, the entropy of both *ought* to converge to around the same value. If you use too few forward passes, the estimates of the mean predictions won’t be very good.

Also, how are the results you are getting at lower step sizes? Do those look alright?

Lewis

**Comment by johannesjung** [comment]
Thank you for the fast response. I've used the default of 100 examples and 20 MC passes. I'll share my results for other configurations.

**Comment by johannesjung** [comment]
I've run some experiments with basic iterativ method, starting with 200 images, now with 1000 to get smoother results. Here the MI ROC curve is constantly the highest in contrast to the previously posted MIM example. Still, the AUC is rather low. I'll also perform some more experiments with MIM, but here are the other results at first. I've used 20 MC samples throughout all experiments.

`config: {'method': 'bim', 'eps': 5, 'clip_min': -103.939, 'clip_max': 131.32, 'ord': inf, 'nb_iter': 10, 'eps_iter': 0.5}`

```
                    AUC entropy AUC bald AUC entropy (succ) AUC bald (succ) Accuracy: 
Deterministic Model 0.3618      0.5      0.3666             0.5             0.424
MC Model            0.2343      0.5153   0.1630             0.4587          0.209
```

**ROC curves**
![image](https://user-images.githubusercontent.com/24568449/40020106-7fcb688a-57c1-11e8-89e1-ca11784a8ade.png)

**PR curves**
![image](https://user-images.githubusercontent.com/24568449/40020112-86171342-57c1-11e8-86fb-69b0989c3a05.png)

Besides those plot I've created some additional ones to get a better intuition about the data.

**Histograms**
![image](https://user-images.githubusercontent.com/24568449/40020229-d80abf28-57c1-11e8-9075-a1c2a40b959f.png)
![image](https://user-images.githubusercontent.com/24568449/40020236-dbf5ca74-57c1-11e8-8220-299cfa20e344.png)
The histograms show why the AUCs are low.

... and for a subset some single examples:
![image](https://user-images.githubusercontent.com/24568449/40020332-15e76a44-57c2-11e8-81b3-3a95c250e1f6.png)

**Comment by lsgos** [comment]
Re your histograms; Yes, the paper mentions this I think; the MI is bounded above by the predictive entropy, and the predictive entropy can't be high when one class is predicted very confidently. Adversarial generation algorithms try to maximise the predictive probability of one class, which means that indirectly they optimise for low entropy. This kind of complicates the picture, as the model uncertainty and the target of adversarial examples are not independent; if the adversarial attacks succeed (i.e they manage to find an example which is classified, say, with probability 0.9999) then the MI based approaches won't help you detect them, since the MI is necessarily low on an image the model classifies with high confidence. (this doesn't help you improve your plots necessarily but I'm guessing you're playing with this code more to understand the paper than to get the exact ROC curve we got). Conversely, having a model which gives you sensible MI estimates is harder to produce an adv example for, whether you actually compute the MI over examples or not.

This is why you can see the phenomenon you observe here, where the ROC is actually _negative_; this reflects behaviour where the model is _more_ confident about adversarial examples than it is about a typical natural image, which is clearly not ideal.

So I think the reason you're having a problem is your trained model. If you are still interested maybe try running it again training for fewer epochs. Looking over my local copies of this file and its history, I think I may only have trained for ~15 epochs when I ran the experiments. 

It's a little bit interesting that training it for too long would reduce the robustness to perturbation, if you find this is the difference; maybe this suggests we should have had higher regularisation on the weight layer.

If this does work I'll update the number of epochs in the repository, since I think this may be a mistake on my part, though I'm a little unsure still when I introduced it.

Thanks for your interest in the paper by the way; sorry the state of the code is a bit less clear than it ought to be
